### PR TITLE
Add ArrowStream source operator

### PR DESCRIPTION
**Pull request by rui-mo**
Add the ArrowStream source operator to read data from an  ArrowArrayStream. 

Fixed: https://github.com/facebookincubator/velox/issues/1254.

### CODE REVIEW

**Review by mbasmanova — COMMENTED** [review]
@rui-mo I wonder if it makes sense to model this logic as a connector, e.g. implement DataSource interface defined in velox/connectors/Connector.h .

**Comment by rui-mo** [comment]
@mbasmanova This operator is used as the source node to fetch data from the arrow array stream. In this way, the arrow data after shuffle is sent to Velox for computing. It can also be feasible to implement it in the connector, we will take a look on that.

**Comment by mbasmanova** [comment]
It is my understanding now that this is an operator that reads shuffle data in Arrow format. If that's the case, then it is fine as a separate operator. (An alternative could be to use the Exchange operator with a custom ExchangeSource.)

Let me read the PR. In the meantime, it would be nice to add a test.

File: velox/core/PlanNode.h [hunk 1]
```diff
@@ -212,6 +214,51 @@ class ValuesNode : public PlanNode {
...
+class ArrowStreamNode : public PlanNode {
+ public:
+  ArrowStreamNode(
+      const PlanNodeId& id,
+      const RowTypePtr& outputType,
+      std::shared_ptr<ArrowArrayStream> arrowStream,
+      bool parallelizable = false)
+      : PlanNode(id),
+        outputType_(outputType),
+        arrowStream_(arrowStream),
```
Comment (mbasmanova) [inline velox/core/PlanNode.h:226]: std::move(arrowStream)

File: velox/core/PlanNode.h [hunk 2]
```diff
@@ -212,6 +214,51 @@ class ValuesNode : public PlanNode {
...
+  }
+
+  const std::vector<PlanNodePtr>& sources() const override;
+
+  bool requiresSplits() const override {
+    return true;
+  }
+
+  // For testing only.
+  bool isParallelizable() const {
```
Comment (mbasmanova) [inline velox/core/PlanNode.h:242]: Is this needed? If not, let's remove.

File: velox/core/PlanNode.h [hunk 3]
```diff
@@ -212,6 +214,51 @@ class ValuesNode : public PlanNode {
...
+    return parallelizable_;
+  }
+
+  std::shared_ptr<ArrowArrayStream> arrowStream() const {
```
Comment (mbasmanova) [inline velox/core/PlanNode.h:246]: use const& for the return type

File: velox/exec/ArrowStream.h [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+#include "velox/vector/arrow/Abi.h"
+
+namespace facebook::velox::exec {
+
+class ArrowStream : public SourceOperator {
+ public:
+  ArrowStream(
+      int32_t operatorId,
+      DriverCtx* driverCtx,
+      std::shared_ptr<const core::ArrowStreamNode> arrowStream);
```
Comment (mbasmanova) [inline velox/exec/ArrowStream.h:28]: const &

File: velox/exec/ArrowStream.h [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+  RowVectorPtr getOutput() override;
+
+  BlockingReason isBlocked(ContinueFuture* /* unused */) override {
+    return BlockingReason::kNotBlocked;
+  }
+
+  void noMoreInput() override {
```
Comment (mbasmanova) [inline velox/exec/ArrowStream.h:36]: Source operators don't receive input. I think this method can be removed.

File: velox/exec/ArrowStream.cpp [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+  rowVector_ = std::dynamic_pointer_cast<RowVector>(
+      facebook::velox::importFromArrowAsViewer(arrowSchema, arrowArray));
+  return rowVector_;
+}
+
+void ArrowStream::close() {
+  closed_ = true;
+}
+
+bool ArrowStream::isFinished() {
```
Comment (mbasmanova) [inline velox/exec/ArrowStream.cpp:53]: isFinished should return true if the operator produced all the data and has no more data to produce. I expect this flag to be set in getOutput(), not in close().

File: velox/exec/ArrowStream.cpp [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+          "Arrow Stream") {
+  arrowStream_ = arrowStream->arrowStream();
+}
+
+RowVectorPtr ArrowStream::getOutput() {
+  struct ArrowArray arrowArray;
+  arrowStream_->get_next(&(*arrowStream_), &arrowArray);
+  if (arrowArray.release == NULL) {
+    // End of Stream.
+    closed_ = true;
```
Comment (mbasmanova) [inline velox/exec/ArrowStream.cpp:38]: closed_ -> finished_ for clarity

File: velox/exec/ArrowStream.cpp [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+  }
+  struct ArrowSchema arrowSchema;
+  arrowStream_->get_schema(&(*arrowStream_), &arrowSchema);
+  // Convert Arrow data into RowVector.
+  rowVector_ = std::dynamic_pointer_cast<RowVector>(
+      facebook::velox::importFromArrowAsViewer(arrowSchema, arrowArray));
+  return rowVector_;
+}
+
+void ArrowStream::close() {
```
Comment (mbasmanova) [inline velox/exec/ArrowStream.cpp:49]: This method can be removed. The base class will take care of close.

File: velox/exec/ArrowStream.cpp [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+  arrowStream_->get_next(&(*arrowStream_), &arrowArray);
+  if (arrowArray.release == NULL) {
+    // End of Stream.
+    closed_ = true;
+    return nullptr;
+  }
+  struct ArrowSchema arrowSchema;
+  arrowStream_->get_schema(&(*arrowStream_), &arrowSchema);
+  // Convert Arrow data into RowVector.
+  rowVector_ = std::dynamic_pointer_cast<RowVector>(
```
Comment (mbasmanova) [inline velox/exec/ArrowStream.cpp:44]: rowVector_ member variable is not needed; Let's remove it and 

```
return std::dynamic_pointer_cast<RowVector>(...);
```

File: velox/exec/ArrowStream.cpp [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+RowVectorPtr ArrowStream::getOutput() {
+  struct ArrowArray arrowArray;
+  arrowStream_->get_next(&(*arrowStream_), &arrowArray);
+  if (arrowArray.release == NULL) {
+    // End of Stream.
+    closed_ = true;
+    return nullptr;
+  }
+  struct ArrowSchema arrowSchema;
+  arrowStream_->get_schema(&(*arrowStream_), &arrowSchema);
```
Comment (mbasmanova) [inline velox/exec/ArrowStream.cpp:42]: Is it necessary to call this for every batch? Can this be called in the constructor?

**Review by mbasmanova — COMMENTED** [review]
@rui-mo Some comments.

File: velox/core/PlanNode.h (hunk 1 above)
Comment (rui-mo) [reply velox/core/PlanNode.h]: Fixed.

File: velox/core/PlanNode.h (hunk 2 above)
Comment (rui-mo) [reply velox/core/PlanNode.h]: Removed.

File: velox/core/PlanNode.h (hunk 3 above)
Comment (rui-mo) [reply velox/core/PlanNode.h]: Fixed.

File: velox/exec/ArrowStream.h (hunk 5 above)
Comment (rui-mo) [reply velox/exec/ArrowStream.h]: Removed.

File: velox/exec/ArrowStream.h (hunk 4 above)
Comment (rui-mo) [reply velox/exec/ArrowStream.h]: Fixed.

File: velox/exec/ArrowStream.cpp (hunk 6 above)
Comment (rui-mo) [reply velox/exec/ArrowStream.cpp:72]: Fixed.

File: velox/exec/ArrowStream.cpp (hunk 7 above)
Comment (rui-mo) [reply velox/exec/ArrowStream.cpp]: Fixed.

File: velox/exec/ArrowStream.cpp (hunk 8 above)
Comment (rui-mo) [reply velox/exec/ArrowStream.cpp:67]: We are overriding this method to call the release of arrow array stream, so still kept this method here.

File: velox/exec/ArrowStream.cpp (hunk 9 above)
Comment (rui-mo) [reply velox/exec/ArrowStream.cpp]: Fixed.

File: velox/exec/ArrowStream.cpp (hunk 10 above)
Comment (rui-mo) [reply velox/exec/ArrowStream.cpp]: Fixed by getting schema in the constructor.

**Comment by rui-mo** [comment]
> It is my understanding now that this is an operator that reads shuffle data in Arrow format. If that's the case, then it is fine as a separate operator. (An alternative could be to use the Exchange operator with a custom ExchangeSource.)
> 
> Let me read the PR. In the meantime, it would be nice to add a test.

Thanks for your comment. Working on the unit test.

File: velox/exec/ArrowStream.cpp [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+  }
+}
+
+RowVectorPtr ArrowStream::getOutput() {
+  struct ArrowArray arrowArray;
+  if (arrowStream_->get_next(arrowStream_.get(), &arrowArray)) {
+    VELOX_FAIL(
+        "Failed to call get_next on ArrowStream: " + std::string(GetError()))
+  }
+  if (arrowArray.release == NULL) {
```
Comment (mbasmanova) [inline velox/exec/ArrowStream.cpp:52]: What is NULL? Can you use nullptr instead?

File: velox/exec/ArrowStream.cpp [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+    finished_ = true;
+    return nullptr;
+  }
+
+  // Convert Arrow Array into RowVector and return.
+  return std::dynamic_pointer_cast<RowVector>(
+      facebook::velox::importFromArrowAsOwner(arrowSchema_, arrowArray, pool_));
+}
+
+const char* ArrowStream::GetError() {
```
Comment (mbasmanova) [inline velox/exec/ArrowStream.cpp:63]: naming: GetError -> getError

File: velox/exec/ArrowStream.h [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+  virtual ~ArrowStream();
+
+  RowVectorPtr getOutput() override;
+
+  BlockingReason isBlocked(ContinueFuture* /* unused */) override {
+    return BlockingReason::kNotBlocked;
+  }
+
+  /// Return last error in array array stream.
+  const char* GetError();
```
Comment (mbasmanova) [inline velox/exec/ArrowStream.h:39]: Should this method be private and const?

File: velox/exec/ArrowStream.h [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+
+  virtual ~ArrowStream();
+
+  RowVectorPtr getOutput() override;
+
+  BlockingReason isBlocked(ContinueFuture* /* unused */) override {
+    return BlockingReason::kNotBlocked;
+  }
+
+  /// Return last error in array array stream.
```
Comment (mbasmanova) [inline velox/exec/ArrowStream.h:38]: typo: "array array stream"

File: velox/exec/ArrowStream.h [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+  const char* GetError();
+
+  bool isFinished() override;
+
+  void close() override;
+
+ private:
+  memory::MemoryPool* pool_;
```
Comment (mbasmanova) [inline velox/exec/ArrowStream.h:46]: not needed; use pool() from base class

File: velox/exec/LocalPlanner.cpp [hunk 16]
```diff
@@ -181,6 +182,11 @@ uint32_t maxDrivers(const DriverFactory& driverFactory) {
       if (!values->isParallelizable()) {
         return 1;
       }
+    } else if (
+        auto arrowStream =
```
Comment (mbasmanova) [inline velox/exec/LocalPlanner.cpp:186]: arrowStream variable is not used, hence, can be dropped

```
if (std::dynamic_pointer_cast<const core::ArrowStreamNode>(node))
```
```

**Review by mbasmanova — COMMENTED** [review]
@rui-mo Rui, looks nice. Some minor comments below. Looking forward to the test.

**Review by mbasmanova — COMMENTED** [review]
Please, add the new plan node and operator to documentation: velox/docs/develop/operators.rst

**Comment by rui-mo** [comment]
This PR is still needed. @zhejiangxiaomai

**Comment by rui-mo** [comment]
This PR is needed but not ready for view because the doc and unit test are not added.

File: velox/exec/ArrowStream.cpp [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+    int32_t operatorId,
+    DriverCtx* driverCtx,
+    const std::shared_ptr<const core::ArrowStreamNode>& arrowStream)
+    : SourceOperator(
+          driverCtx,
+          arrowStream->outputType(),
+          operatorId,
+          arrowStream->id(),
+          "Arrow Stream") {
+  pool_ = arrowStream->memoryPool();
```
[the remark this answers is not in the document]
Comment (rui-mo) [reply velox/exec/ArrowStream.cpp]: Fixed.

File: velox/exec/ArrowStream.cpp (hunk 11 above)
Comment (rui-mo) [reply velox/exec/ArrowStream.cpp:51]: Arrow marks it as NULL in release callback, so we follow that here. 
https://github.com/apache/arrow/blob/master/cpp/src/arrow/c/helpers.h#L63-L64

File: velox/exec/ArrowStream.cpp (hunk 12 above)
Comment (rui-mo) [reply velox/exec/ArrowStream.cpp]: Fixed.

File: velox/exec/ArrowStream.h (hunk 13 above)
Comment (rui-mo) [reply velox/exec/ArrowStream.h]: Fixed.

File: velox/exec/ArrowStream.h (hunk 14 above)
Comment (rui-mo) [reply velox/exec/ArrowStream.h]: Fixed.

File: velox/exec/ArrowStream.h (hunk 15 above)
Comment (rui-mo) [reply velox/exec/ArrowStream.h:48]: Yes, fixed.

File: velox/exec/LocalPlanner.cpp (hunk 16 above)
Comment (rui-mo) [reply velox/exec/LocalPlanner.cpp]: Fixed.

**Comment by rui-mo** [comment]
Hi @mbasmanova @majetideepak, it has been a long time and sorry for the late response because we were on other tasks. This PR mainly supports an operation to read data from arrow array stream, and I fixed above comments including adding unit test and updating the doc.
Is it possible to reopen this PR, or do I need to open a new one for your review? My changes are still in this branch: https://github.com/rui-mo/velox/commits/wip_arrow_stream. Thanks for your help!

**Comment by mbasmanova** [comment]
@rui-mo It should be possible to re-open the PR, but I see that "Reopen and comment" button is greyed out. Would you try updating the associated branch with the latest code? CC:@kgpai

**Comment by rui-mo** [comment]
> @rui-mo It should be possible to re-open the PR, but I see that "Reopen and comment" button is greyed out. Would you try updating the associated branch with the latest code? CC:@kgpai

@mbasmanova Thanks for your advice. Updated the associated branch to the latest status, but the PR content did not change. I'm glad to open a new one if it's needed.

**Comment by mbasmanova** [comment]
@rui-mo Let's open a new PR.

**Comment by rui-mo** [comment]
@mbasmanova Hi Masha, a new PR https://github.com/facebookincubator/velox/pull/3917 was opened to replace this one. Would you like to take a review? Thanks.

**Comment by mbasmanova** [comment]
@rui-mo Thank you, Rui. I have a few things I need to work on today and tomorrow. I'll try to get to this PR early next week.

**Comment by rui-mo** [comment]
> @rui-mo Thank you, Rui. I have a few things I need to work on today and tomorrow. I'll try to get to this PR early next week.

@mbasmanova Sure, thanks for your time!

**Pull request closed without merging**